### PR TITLE
Make linear arithmetic use its inference manager

### PR DESCRIPTION
**Pull request by nafur**
This PR refactors the linear arithmetic solver to properly use its inference manager, instead of directly sending lemmas to the output channel. To do this, it introduces new InferenceIds for the various linear lemmas.

### CODE REVIEW

File: src/theory/theory_inference_manager.h [hunk 1]
```diff
@@ -135,12 +135,12 @@ class TheoryInferenceManager
    * Raise conflict conf (of any form), without proofs. This method should
    * only be called if there is not yet proof support in the given theory.
    */
-  void conflict(TNode conf, InferenceId id);
+  void conflict(TNode conf, InferenceId id, bool addAlways = false);
```
Comment (ajreynol) [inline src/theory/theory_inference_manager.h:138]: I think I'd prefer not adding this override.

**Comment by ajreynol** [comment]
At a high level, I'd prefer not adding the override for now.

We should test this on SMT-LIB.

Discussing with Gereon offline, the dilemma is that linear arithmetic doesnt generally cache lemmas but non-linear does, but they use a common inference manager.  My recommendation would be to enable lemma caching when non-linear is enabled only, which would mean the change of this PR would be to cache also the linear lemmas when non-linear is enabled.

**Comment by nafur** [comment]
The results of this PR are surprisingly nice.
Non-incremental:
```
                                  status  total  solved  sat  unsat  best  timeout  memout  error  uniq  disagr  time_cpu  memory
benchmark config                                                                                                                 
AUFLIRA   03040701-master-noninc/     ok      6       4    0      4     4        2       0      0     4       0    1076.6   220.8
          03040701-pr-noninc/         ok      6       2    0      2     2        4       0      0     2       0    1570.1   221.8
AUFNIRA   03040701-master-noninc/     ok      1       1    0      1     1        0       0      0     1       0     275.8   213.0
          03040701-pr-noninc/         to      1       0    0      0     0        1       0      0     0       0     300.3   188.1
QF_ABV    03040701-master-noninc/     to      2       0    0      0     0        2       0      0     0       0     600.6   514.3
          03040701-pr-noninc/         ok      2       2    2      0     2        0       0      0     2       0     596.2   517.8
QF_ALIA   03040701-master-noninc/     ok      3       2    0      2     2        1       0      0     2       0     781.1   394.6
          03040701-pr-noninc/         ok      3       1    1      0     1        2       0      0     1       0     857.5   421.5
QF_BV     03040701-master-noninc/     ok      8       5    2      3     5        3       0      0     5       0    2380.8  5275.4
          03040701-pr-noninc/         ok      8       3    1      2     3        5       0      0     3       0    2395.6  5248.7
QF_BVFP   03040701-master-noninc/     to      1       0    0      0     0        1       0      0     0       0     300.3  1147.6
          03040701-pr-noninc/         ok      1       1    0      1     1        0       0      0     1       0     253.1  1162.2
QF_FP     03040701-master-noninc/     to      1       0    0      0     0        1       0      0     0       0     300.4  1079.8
          03040701-pr-noninc/         ok      1       1    0      1     1        0       0      0     1       0     295.8  1109.8
QF_IDL    03040701-master-noninc/     ok      2       1    0      1     1        1       0      0     1       0     598.4   688.9
          03040701-pr-noninc/         ok      2       1    0      1     1        1       0      0     1       0     594.1   684.5
QF_LIA    03040701-master-noninc/     ok      9       4    1      3     4        5       0      0     4       0    2671.3  4555.4
          03040701-pr-noninc/         ok      9       5    2      3     5        4       0      0     5       0    2698.6  4603.5
QF_LRA    03040701-master-noninc/     to      1       0    0      0     0        1       0      0     0       0     300.3   275.9
          03040701-pr-noninc/         ok      1       1    1      0     1        0       0      0     1       0     298.3   274.8
QF_NIA    03040701-master-noninc/     ok    519       6    6      0     6      508       0      0     6       0  154089.3 24001.9
          03040701-pr-noninc/         ok    519     513  141    372   513        4       0      0   513       0   26203.2 32019.4
QF_NRA    03040701-master-noninc/     to      3       0    0      0     0        3       0      0     0       0     900.9  1124.6
          03040701-pr-noninc/         ok      3       3    0      3     3        0       0      0     3       0     826.3  1135.1
QF_SLIA   03040701-master-noninc/     ok      9       4    3      1     4        5       0      0     4       0    2002.7   534.8
          03040701-pr-noninc/         ok      9       5    3      2     5        4       0      0     5       0    2071.1   578.1
QF_UF     03040701-master-noninc/     ok      1       1    0      1     1        0       0      0     1       0     288.2    84.0
          03040701-pr-noninc/         to      1       0    0      0     0        1       0      0     0       0     300.3    82.9
UFDTLIRA  03040701-master-noninc/     to      1       0    0      0     0        1       0      0     0       0     300.3   128.4
          03040701-pr-noninc/         ok      1       1    0      1     1        0       0      0     1       0     120.6    85.1
UFDTNIRA  03040701-master-noninc/     to      3       0    0      0     0        3       0      0     0       0     900.9    39.0
          03040701-pr-noninc/         ok      3       3    0      3     3        0       0      0     3       0      63.2  2429.9
UFLIA     03040701-master-noninc/     to      1       0    0      0     0        1       0      0     0       0     300.7  6447.4
          03040701-pr-noninc/         ok      1       1    0      1     1        0       0      0     1       0     297.3  6448.4
UFNIA     03040701-master-noninc/     ok    105       2    0      2     2      102       0      0     2       0   31390.3 12465.9
          03040701-pr-noninc/         ok    105     103   11     92   103        2       0      0   103       0    5149.0 15181.7
```
Incremental:
```
                                 status  total   solved      sat   unsat  best  timeout  memout  error  uniq  disagr  time_cpu    memory
benchmark   config                                                                                                                      
ABVFP       03040701-master-inc/     ok      4      818      801      17     3        1       0      0     0       0     311.7     414.1
            03040701-pr-inc/         ok      4      818      801      17     3        1       0      0     0       0     311.5     411.5
ALIA        03040701-master-inc/     ok     24   189307   181077    8230     5       16       0      0     0       0    4924.1    2680.8
            03040701-pr-inc/         ok     24   202552   193238    9314    21        0       0      0     0       0     640.1    2573.8
ANIA        03040701-master-inc/     to      3      509      479      30     0        3       0      0     0       0     900.9     148.7
            03040701-pr-inc/         ok      3    23488    22768     720     3        0       0      0     0       0      70.1     163.7
AUFNIRA     03040701-master-inc/     ok    136     3085        0    3085   122        0       0      0     0       0     902.9    2160.7
            03040701-pr-inc/         ok    136     3085        0    3085   105        0       0      0     0       0     906.6    2155.3
BV          03040701-master-inc/     ok     18    35857    32486    3371    12        7       0      0     0       0    2924.8    8001.4
            03040701-pr-inc/         ok     18    35859    32488    3371    13        7       0      0     0       0    2928.6    7927.6
BVFP        03040701-master-inc/     ok     10     3422     3062     360     9        6       0      0     0       0    1853.9    1533.3
            03040701-pr-inc/         ok     10     3422     3062     360     9        6       0      0     0       0    1854.8    1536.1
LIA         03040701-master-inc/     ok      6    25393    21716    3677     3        0       0      0     0       0      23.2     139.5
            03040701-pr-inc/         ok      6    25393    21716    3677     5        0       0      0     0       0      23.0     139.1
LRA         03040701-master-inc/     ok      5    15969    15901      68     5        0       0      0     0       0      79.0     575.7
            03040701-pr-inc/         ok      5    15969    15901      68     2        0       0      0     0       0      80.9     576.6
QF_ABV      03040701-master-inc/     ok    948     2251      959    1292   710       22       0      0     0       0   18137.9   94315.6
            03040701-pr-inc/         ok    948     2247      957    1290   695       24       0      0     0       0   18140.3   94377.6
QF_ABVFP    03040701-master-inc/     ok     14     1960     1868      92    11        1       0      0     0       0     622.0    3903.7
            03040701-pr-inc/         ok     14     1960     1868      92    12        1       0      0     0       0     631.3    3931.0
QF_ALIA     03040701-master-inc/     ok     44   398938   373392   25546    19       19       0      0     0       0    6164.7   10738.0
            03040701-pr-inc/         ok     44   467906   435320   32586    32        5       0      0     0       0    2592.2   10751.6
QF_ANIA     03040701-master-inc/     to      5      927      867      60     0        5       0      0     0       0    1501.5     288.6
            03040701-pr-inc/         ok      5    51724    49889    1835     5        0       0      0     0       0     233.4     382.4
QF_AUFBV    03040701-master-inc/     ok     30      496        7     489    26       17       0      0     0       0    5405.9   11662.0
            03040701-pr-inc/         ok     30      495        7     488    22       17       0      0     0       0    5408.2   11660.0
QF_AUFBVLIA 03040701-master-inc/     ok    345      690      349     341   216        0       0      0     0       0   11204.7   80655.1
            03040701-pr-inc/         ok    345      690      349     341   222        0       0      0     0       0   11212.0   80650.2
QF_AUFBVNIA 03040701-master-inc/     ok     30       60       32      28    16        0       0      0     0       0     986.7    9175.8
            03040701-pr-inc/         ok     30       59       31      28    20        1       0      0     0       0     986.6    9183.2
QF_AUFLIA   03040701-master-inc/     ok     72  1875479  1463797  411682    50       23       0      0     0       0    8104.1   23587.5
            03040701-pr-inc/         ok     72  1874019  1462281  411738    49       23       0      0     0       0    8073.7   22910.3
QF_BV       03040701-master-inc/     ok   1205    49672    43842    5830  1009       41       0      0     0       0   15986.2  111760.7
            03040701-pr-inc/         ok   1205    49671    43843    5828   965       39       0      0     2       0   16009.5  111665.9
QF_BVFP     03040701-master-inc/     ok    114    32839    32167     672    96        0       0      0     0       0     517.2    9092.7
            03040701-pr-inc/         ok    114    32839    32167     672    81        0       0      0     0       0     516.6    9087.6
QF_FP       03040701-master-inc/     ok      2       52       41      11     2        0       0      0     0       0       0.5      17.8
            03040701-pr-inc/         ok      2       52       41      11     2        0       0      0     0       0       0.5      16.1
QF_LIA      03040701-master-inc/     ok     69  2888432  2227999  660433    43       55       0      0     0       0   17254.7   35747.7
            03040701-pr-inc/         ok     69  2905145  2239812  665333    45       55       0      0     0       0   17243.2   34802.2
QF_LRA      03040701-master-inc/     to     10      248       48     200     4       10       0      0     0       0    3004.0    5893.9
            03040701-pr-inc/         to     10      248       48     200     9       10       0      0     0       0    3002.0    5635.9
QF_NIA      03040701-master-inc/     to     10   390305   363129   27176     1       10       0      0     0       0    3003.1    4675.3
            03040701-pr-inc/         ok     10  1659619  1614612   45007    10        4       0      0     0       0    2137.4    4625.0
QF_UF       03040701-master-inc/     ok    764     1528     1066     462   633        0       0      0     0       0     986.2   43271.4
            03040701-pr-inc/         ok    764     1528     1066     462   558        0       0      0     0       0     998.7   43269.8
QF_UFBV     03040701-master-inc/     ok   2323     5049     3332    1717  1672       10       0      0     0       0   18851.6  172124.7
            03040701-pr-inc/         ok   2323     5048     3332    1716  1529       10       0      0     0       0   19027.0  172122.9
QF_UFBVLIA  03040701-master-inc/     ok     28       56       45      11    21        0       0      0     0       0       4.8     298.3
            03040701-pr-inc/         ok     28       56       45      11    23        0       0      0     0       0       4.7     302.3
QF_UFFP     03040701-master-inc/     ok      1        2        0       2     1        0       0      0     0       0       0.5      45.5
            03040701-pr-inc/         ok      1        2        0       2     1        0       0      0     0       0       0.5      32.0
QF_UFLIA    03040701-master-inc/     ok    603   288163   274336   13827   560      219       0      0     0       0   66411.7   20739.9
            03040701-pr-inc/         ok    603   444271   421779   22492   533      214       0      0     0       0   65094.2   21440.8
QF_UFLRA    03040701-master-inc/     ok   3033    19606    12150    7456  2564      199       0      0     0       0   82842.0  257676.9
            03040701-pr-inc/         ok   3033    19608    12154    7454  2537      200       0      0     0       0   83071.5  256957.9
QF_UFNIA    03040701-master-inc/     to      1    22186    20991    1195     0        1       0      0     0       0     300.3     157.5
            03040701-pr-inc/         ok      1    23607    22195    1412     1        0       0      0     0       0      33.6     146.9
UF          03040701-master-inc/     ok   2954    75844    32924   42920  2404      538      16      0     0       0  182861.5 1471812.8
            03040701-pr-inc/         ok   2954    75868    32924   42944  2433      535      18      0     0       0  182807.7 1470982.7
UFLRA       03040701-master-inc/     ok   1038   237898        0  237898   739       42       0      0     0       0   36842.5   24848.7
            03040701-pr-inc/         ok   1038   237849        0  237849   556       42       0      0     0       0   37093.0   24832.0
UFNIA       03040701-master-inc/     ok   3160    94702    30165   64537  2546      659       0      0     0       0  220792.4  855399.6
            03040701-pr-inc/         ok   3160    96352    30419   65933  2557      635       0      0     0       0  214995.1  855292.6
UFNRA       03040701-master-inc/     ok      3       10        6       4     3        0       0      0     0       0       1.2      13.6
            03040701-pr-inc/         ok      3       10        6       4     2        0       0      0     0       0       1.3      13.8
```

**Review by aniemetz — COMMENTED** [review]
I'd prefer having these tests fixed rather than removed, if possible. Discussing this with @nafur and @ajreynol  offline.

**Comment by nafur** [comment]
> I'd prefer having these tests fixed rather than removed, if possible. Discussing this with @nafur and @ajreynol offline.

After a lot of back and forth, we settled for removing these test cases. They make extremely strong assumptions and fix the theory behavior way too strictly. We also remove `Theory::setOutputChannel` which only existed for this particular test case.

**Review by ajreynol — APPROVED** [review]
LGTM, I'm excited about these results.  Thanks for all the SMT-LIB runs.

**Pull request merged**